gstd/src/common/reservations.rs

```diff
@@ -0,0 +1,328 @@
+
+        let insert_to = self.0[insert_range_start..insert_range_end]
+            .binary_search_by(|reservation| reservation.valid_until.cmp(&valid_until));
+        let insert_to = if insert_range_start == self.0.len() {
+            self.0.len()
+        } else {
+            match insert_to {
+                Ok(pos) => pos + 1,
+                Err(pos) => pos,
+            }
+        };
+
+        // self.0 will allocate new memory.
+        if self.0.capacity() == self.0.len() {
+            self.cleanup(block_height);
+        }
+
+        self.0.insert(
+            insert_to,
+            Reservation {
+                id,
+                amount,
+                valid_until,
+            },
+        );
+
+        Ok(())
+    }
+
+    /// Find the appropriate reservation with reserved amount greater than or
+    /// equal to `amount`.
+    ///
+    /// If such a reservation is found, [`ReservationId`] is returned.
+    ///
+    /// # Underlying logics
+    ///
+    /// Executes for O(logN)..O(N), where N is amount of stored
+    /// reservations. When there's many expired reservations execution time
+    /// is closer to O(N).
+    ///
+    /// All the reservations are sorted by their amount and then by the time
+    /// when they'll expire (both are in ascending order) in underlying `Vec`,
+    /// so when one's trying to take reservation, reservation with the least
+    /// possible amount is found and if it's already expired, the search to the
+    /// end of underlying `Vec` occurs. After that, all the expired
+    /// reservations that were found in process of search are cleaned out.
+    ///
+    /// # See also
+    /// - [`ReservationId`] is used to reserve and unreserve gas amount
+    /// for program execution later.
+    pub fn try_take_reservation(&mut self, amount: u64) -> Option<ReservationId> {
+        let search_from = self
+            .0
+            .partition_point(|reservation| reservation.amount < amount);
+
+        if search_from < self.0.len() {
+            let block_height = exec::block_height();
+            for i in search_from..self.0.len() {
+                if self.0[i].valid_until > block_height {
+                    // All the checked reservations are already expired at this time.
+                    let suitable = self
+                        .0
+                        .drain(search_from..=i)
+                        .last()
+                        .expect("At least one element in range");
+
+                    return Some(suitable.id);
+                }
+            }
+        }
+
+        // All the checked reservations are already expired at this time.
+        self.0.drain(search_from..);
+
+        None
+    }
+
+    /// Returns an amount of the stored reservations
+    /// that aren't expired at this time.
+    ///
+    /// Executes for O(N) where N is amount of stored reservations.
+    pub fn count_valid(&self) -> usize {
```

Comment:
Do we really need this?

Reply:
I guess business logic can rely on valid amount of reservations. There is no way to count valid ones from the API user level.